muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/CommunicatorController.java

```diff
@@ -350,11 +355,50 @@ public CommunicatorMessageRecipient findCommunicatorMessageRecipientByMessageAnd
   }
     
   public List<CommunicatorMessageRecipientUserGroup> listCommunicatorMessageUserGroupRecipients(CommunicatorMessage communicatorMessage) {
-    return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    UserEntity loggedUser = sessionController.getLoggedUserEntity();
+    UserEntity sender = userEntityController.findUserEntityById(communicatorMessage.getSender());
+    if (sessionController.hasEnvironmentPermission(CommunicatorPermissionCollection.COMMUNICATOR_READ_RECIPIENTS_LIST)) {
+      return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    } else if (loggedUser.defaultSchoolDataIdentifier().equals(sender.defaultSchoolDataIdentifier())) {
+      return communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+    } else {
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipientsList = communicatorMessageRecipientUserGroupDAO.listByMessage(communicatorMessage);
+      List<CommunicatorMessageRecipientUserGroup> userGroupRecipients = new ArrayList<CommunicatorMessageRecipientUserGroup>();
+      SchoolDataIdentifier loggedUserIdentifier = sessionController.getLoggedUserEntity().defaultSchoolDataIdentifier();
+  
+      for (CommunicatorMessageRecipientUserGroup userGroup : userGroupRecipientsList) {
+        UserGroupEntity userGroupEntity = userGroupEntityController.findUserGroupEntityById(userGroup.getUserGroupEntityId());
+        List<UserGroupUserEntity> groupUsers = userGroupEntityController.listUserGroupUserEntitiesByUserGroupEntity(userGroupEntity);
```

Comment:
Might be faster using `userGroupEntityController.isMember(loggedUserIdentifier, userGroupEntity)`

Reply:
Seconded, the `groupUsers.contains(groupUserEntity)` below can be bit iffy as it's going to use `object.equals`, which I wouldn't count on when the objects are Entities that don't have explicit compare functions.